libkrun/libkruntime.c

```diff
@@ -37,86 +79,254 @@ clock_gettime_monotonic()
-        : "%rax", "%rbx", "%rcx", "%rdx");  // other clobbers
-    return ((u_int64_t) hi << 32) | (u_int64_t) lo;
+    asm volatile(
+        "mov $0xa, %%eax\n\t"       // pctr leaf
+        "cpuid\n\t"
+        : "=a" (eax), "=d" (edx)    // out
+        :                           // in
+        : "ebx", "ecx");            // clobber
+
+    /* eax
+     *   0-4:  number of fixed-func counters
+     *   5-12: width of counters
+     */
+    num_fixed_ctrs = edx & 0x1f;
+    fixed_ctr_width = (edx & 0x1fe0) >> 5;
+
+    /* eax
+     *   0-7:  architectural counter version
+     *   8-31: reserved
+     */
+    arch_ctr_vers = eax & (0xff);
+
+    // fixed function perf ctrs appeared on arch counter version 2
+    if (arch_ctr_vers < 2) {
+        printf("arch pctr version >=2 is required! got %d\n", arch_ctr_vers);
+        exit(EXIT_FAILURE);
+    }
+
+    // We are interested in IA32_FIXED_CTR1, (i.e. the second fixed counter)
+    if (num_fixed_ctrs < 2) {
+        fprintf(stderr, "too few fixed-function counters: %d\n",
+            num_fixed_ctrs);
+        exit(EXIT_FAILURE);
+    }
+
+    return fixed_ctr_width;
+}
+#endif // linux && !TRAVIS
+
+void
+libkruntime_init(void)
+{
+#if defined(__linux__) && !defined(TRAVIS)
+    int core;
+
+    /* See how wide the counter values are and make an appropriate mask */
+    pctr_val_mask = ((uint64_t) 1 << get_fixed_pctr1_width()) - 1;
+
+    /* Set up MSR device node handles */
+    num_msr_nodes = sysconf(_SC_NPROCESSORS_ONLN);
+
+    msr_r_nodes = calloc(num_msr_nodes, sizeof(int));
+    if (msr_r_nodes == NULL) {
+        perror("calloc");
+        exit(EXIT_FAILURE);
+    }
+
+    msr_w_nodes = calloc(num_msr_nodes, sizeof(int));
+    if (msr_w_nodes == NULL) {
+        perror("calloc");
+        exit(EXIT_FAILURE);
+    }
+
+    for (core = 0; core < num_msr_nodes; core++) {
+        msr_r_nodes[core] = open_msr_node(core, O_RDONLY);
+        msr_w_nodes[core] = open_msr_node(core, O_WRONLY);
+    }
+
+    /* Configure and reset CPU_CLK_UNHALTED.CORE on all CPUs */
+    config_fixed_ctr1_all_cores(1);
+#else
+    /*
+     * Does nothing. rmsr device node is Linux specific and travis does not
+     * have performance counters.
+     */
+#endif  // __linux__ && !TRAVIS
+}
+
+void
+libkruntime_done(void)
+{
+#if defined(__linux__) && !defined(TRAVIS)
+    int core;
+
+    /* Close MSR device nodes */
+    for (core = 0; core < num_msr_nodes; core++) {
+        close_fd(msr_r_nodes[core]);
+        close_fd(msr_w_nodes[core]);
+    }
+#else
+    // Nothing
+#endif  // __linux__ && !TRAVIS
 }
 
 u_int64_t
-read_ts_reg_stop()
+read_core_cycles(void)
 {
-    u_int32_t hi, lo;
+#if defined(__linux__) && !defined(TRAVIS)
+    int core;
+    uint64_t cycles = 0, val;
+
+    for (core = 0; core < num_msr_nodes; core++) {
+        val = read_msr(core, PCTR_IA32_PERF_FIXED_CTR1) & pctr_val_mask;
+
+        // Note that overflow is impossible on most platforms. E.g. on an i7,
```

Comment:
Even for a really long-running benchmark like CPython? If that's the case then the assert is pretty pointless.